src/charm.py

```diff
@@ -979,6 +1057,118 @@ def auth_enabled(self) -> bool:
 
         return any("MONGOD_ARGS" in line and "--auth" in line for line in env_vars)
 
+    def _scope_obj(self, scope: Scopes):
+        if scope == APP_SCOPE:
+            return self.app
+        if scope == UNIT_SCOPE:
+            return self.unit
+
+    def _peer_data(self, scope: Scopes):
+        if not self._peers:
+            return {}.setdefault(scope, {})
+        scope_obj = self._scope_obj(scope)
+        return self._peers.data[scope_obj]
+
+    def _juju_secret_set(self, scope: Scopes, key: str, value: str) -> str:
+        """Helper function setting Juju secret."""
+        peer_data = self._peer_data(scope)
+        self._juju_secrets_get(scope)
+
+        secret = self.secrets[scope].get(Config.Secrets.SECRET_LABEL)
+
+        # It's not the first secret for the scope, we can re-use the existing one
+        # that was fetched in the previous call
+        if secret:
+            secret_cache = self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL]
+
+            if secret_cache.get(key) == value:
+                logging.debug(f"Key {scope}:{key} has this value defined already")
+            else:
+                secret_cache[key] = value
+                try:
+                    secret.set_content(secret_cache)
+                    logging.debug(f"Secret {scope}:{key} was {key} set")
+                except OSError as error:
+                    logging.error(
+                        f"Error in attempt to set '{key}' secret for scope '{scope}'. "
+                        f"Existing keys were: {list(secret_cache.keys())}. {error}"
+                    )
+
+        # We need to create a brand-new secret for this scope
+        else:
+            scope_obj = self._scope_obj(scope)
+
+            secret = scope_obj.add_secret({key: value})
+            if not secret:
+                raise SecretNotAddedError(f"Couldn't set secret {scope}:{key}")
+
+            self.secrets[scope][Config.Secrets.SECRET_LABEL] = secret
+            self.secrets[scope][Config.Secrets.SECRET_CACHE_LABEL] = {key: value}
+            logging.debug(f"Secret {scope}:{key} published (as first). ID: {secret.id}")
+            peer_data.update({Config.Secrets.SECRET_INTERNAL_LABEL: secret.id})
+
+        return self.secrets[scope][Config.Secrets.SECRET_LABEL].id
+
+    def _juju_secrets_get(self, scope: Scopes) -> Optional[bool]:
+        """Helper function to get Juju secret."""
+        peer_data = self._peer_data(scope)
+
+        if not peer_data.get(Config.Secrets.SECRET_INTERNAL_LABEL):
```

Comment:
noob question, why is it needed to check for internal secrets in peer data? is it because it means that the charm hasn't been set up, i.e. mongod user etc? (maybe adding a comment here will be useful for maintainability)

Reply:
@juditnovak ^^ 😄

Reply:
Peer data bag stores the Secret ID inside in `SECRET_INTERNAL_LABEL` field.
The current code cannot retrieve the Secret object if we don't have the Secret ID (line 1126 below).
I think this part is going to be revoked in two weeks.

Reply:
We store the ID of the the secret object in the peer data.
If the ID is not there, we can't fetch the secret (yet).

So technically this means that no secret was set up for Mongo yet.
If I'm correct it's the username/password that comes first, basically initiating the secret.

However not that this will change in a week or two -- we'll be using Secret labels instead of IDs in Mongo.